src/PowerShellEditorServices/Services/PowerShell/Host/InternalHost.cs

```diff
@@ -0,0 +1,778 @@
+        {
+            while (!_shouldExit)
+            {
+                using (CancellationScope cancellationScope = _cancellationContext.EnterScope(isIdleScope: false))
+                {
+                    DoOneRepl(cancellationScope.CancellationToken);
+
+                    if (_shouldExit)
+                    {
+                        break;
+                    }
+
+                    while (!cancellationScope.CancellationToken.IsCancellationRequested
+                        && _taskQueue.TryTake(out ISynchronousTask task))
+                    {
+                        task.ExecuteSynchronously(cancellationScope.CancellationToken);
+                    }
+                }
+            }
+        }
+
+        private void DoOneRepl(CancellationToken cancellationToken)
+        {
+            if (_skipNextPrompt)
+            {
+                _skipNextPrompt = false;
+                return;
+            }
+
+            try
+            {
+                string prompt = GetPrompt(cancellationToken);
+                UI.Write(prompt);
+                string userInput = InvokeReadLine(cancellationToken);
+
+                // If the user input was empty it's because:
+                //  - the user provided no input
+                //  - the readline task was canceled
+                //  - CtrlC was sent to readline (which does not propagate a cancellation)
+                //
+                // In any event there's nothing to run in PowerShell, so we just loop back to the prompt again.
+                // However, we must distinguish the last two scenarios, since PSRL will not print a new line in those cases.
+                if (string.IsNullOrEmpty(userInput))
+                {
+                    if (cancellationToken.IsCancellationRequested
+                        || LastKeyWasCtrlC())
+                    {
+                        UI.WriteLine();
+                    }
+                    return;
+                }
+
+                InvokeInput(userInput, cancellationToken);
+            }
+            catch (OperationCanceledException)
+            {
+                // Do nothing, since we were just cancelled
+            }
+            catch (Exception e)
+            {
+                UI.WriteErrorLine($"An error occurred while running the REPL loop:{Environment.NewLine}{e}");
+                _logger.LogError(e, "An error occurred while running the REPL loop");
+            }
+        }
+
+        private string GetPrompt(CancellationToken cancellationToken)
+        {
+            var command = new PSCommand().AddCommand("prompt");
+            IReadOnlyList<string> results = InvokePSCommand<string>(command, PowerShellExecutionOptions.Default, cancellationToken);
+            string prompt = results.Count > 0 ? results[0] : DefaultPrompt;
+
+            if (CurrentRunspace.RunspaceOrigin != RunspaceOrigin.Local)
+            {
+                prompt = Runspace.GetRemotePrompt(prompt);
+            }
```

Comment:
We steal this from PowerShell to get the nice remote prompt customisation. It's something that PowerShell should just expose publicly